src/Crypto/WebAuthn/Operation/Authentication.hs

```diff
@@ -160,26 +157,28 @@ newtype AuthenticationResult = AuthenticationResult
 -- The 'arSignatureCounterResult' field of the result should be inspected to
 -- enforce Relying Party policy regarding potentially cloned authenticators.
 verifyAuthenticationResponse ::
+  (Monad m) =>
   -- | The origin of the server
   M.Origin ->
   -- | The hash of the relying party id
   M.RpIdHash ->
-  -- | The user handle, in case the user is identified already
-  -- TODO: Mention that this would be empty for username-less authentication
+  -- | The currently identified user. Set if the user was identified  before
+  -- the ceremony was initiated, e.g., via a username or cookie.
+  -- This should be set to 'Nothing' if you are implementing a login flow where
+  -- passkeys are automatically filled in by the browser in the login form.
   Maybe M.UserHandle ->
-  -- | The database entry for the credential, as created in the initial
-  -- attestation and optionally updated in subsequent assertions
-  CredentialEntry ->
+  -- | Function to look up the credential in the database. This function is used
+  -- to verify that the credential belongs to the user identified by either the
+  -- 'userIdentified' parameter if it is not 'Nothinbg' or the 'M.araUserHandle'
+  -- field of the response.
+  (M.UserHandle -> M.CredentialId -> m (Maybe CredentialEntry)) ->
```

Comment:
This corresponds to a database lookup `SELECT credential_id,public_key FROM credential_entries where handle = ?`

Reply:
I think it's unfortunate that we're switching to trusting the consumer of this API to validate that the UserHandle and CredentialId correspond.

If I were a consumer of this api, my instinct would be to just supply

`lookupCredential credentialId _ = getBy $ UniqueWebAuthnCredentialId credentialId`

which would potentially introduce a security vulnerability.

The return value of `Nothing` is doing double-duty of "I don't recognize this credential id", and "this credential id is invalid for this user".

Instead, I would propose keeping `ceUserHandle` on `CredentialEntry` so that the library can handle this check, and either:

a. Changing the signature of this method to `(M.CredentialId -> m (Maybe CredentialEntry))`
b. I'd take a serious look at whether we need this callback, or if we can just help make it convenient for the consumer to extract `CredentialId` from `Credential` so that they can supply the correct `CredentialEntry` when calling the function.

I'm new to the code, but (b) _feels_ like the right move to me.
1. There's a weak argument around complexity. The consumer has to supply a function that can produce a `CredentialEntry` for any `CredentialId`, vs just a single `CredentialEntry`. 
2. It seems like it'd be really easy to implement (b). You wouldn't need to make any changes to this function's API, so you wouldn't need to fix the `server` stuff, and we wouldn't need to make any changes to the existing webauthn BE code. If `Crypto.WebAuthn.Model` is exported from this package, there might not actually be _any_ changes required to this package to get us a functional solution.

Reply:
I generally agree with @m5 that it's unfortunate to have a callback and change this from a pure to a lifted function.

If @m5's suggestions work I think we could do that. If this `lookupCredential` is really necessary to have in the middle of the flow, could we consider splitting this into two separate pure functions, with the first one returning both `M.UserHandle`, `M.CredentialId` and any data necessary to call the second one?

Reply:
that makes sense!

I guess we kinda sorta already have to be  doing`b` in mwb today right?    Because we somehow need to pass in that `CredentialEntry` and I hope we look that up by  the tuple `(credentialId, userHandleFromSession)`.   So we are already extracting things from the received `Credential` and doing a database lookup before calling `verifyAuthenticationResponse`

I think you've convinced me we can implement this without changes to the library.